reframe/listener.go

```diff
@@ -0,0 +1,507 @@
+	dsWrapper    *dsWrapper
+	engine       provider.Interface
+	cidTtl       time.Duration
+	chunkSize    int
+	currentChunk *cidsChunk
+	// ReframeListener maintains in memory indexes for fast key value lookups
+	// as well as a rolling double-linked list of CIDs ordered by their timestamp.
+	// Once a CID gets advertised, the respective linked list node gets moved to the
+	// beginning of the list. To identify CIDs to expire, ReframeListener would walk the list tail to head.
+	// TODO: offload cid chunks to disk to save RAM
+	chunkByContextId       map[string]*cidsChunk
+	chunkByCid             map[cid.Cid]*cidsChunk
+	listNodeByCid          map[cid.Cid]*list.Element
+	nodesLl                *list.List
+	nonceGen               func() []byte
+	lastSeenProviderInfo   *peer.AddrInfo
+	configuredProviderInfo *peer.AddrInfo
+	stats                  stats
+}
+
+type stats struct {
+	putAdsSent            int
+	removeAdsSent         int
+	cidsProcessed         int
+	existingCidsProcessed int
+	cidsExpired           int
+	reframeCallsReceived  int
+}
+
+type cidNode struct {
+	timestamp time.Time
+	c         cid.Cid
+}
+type cidsChunk struct {
+	ContextID []byte
+	Cids      map[cid.Cid]struct{}
+}
+
+func defaultNonceGen() []byte {
+	nonce := make([]byte, 8)
+	rand.Read(nonce)
+	return nonce
+}
+
+type ReframeMultihashLister struct {
+	CidFetcher func(contextID []byte) (map[cid.Cid]struct{}, error)
+}
+
+func (lister *ReframeMultihashLister) MultihashLister(ctx context.Context, p peer.ID, contextID []byte) (provider.MultihashIterator, error) {
+	contextIdStr := contextIDToStr(contextID)
+	cids, err := lister.CidFetcher(contextID)
+
+	if err != nil {
+		return nil, err
+	}
+
+	mhs := make([]multihash.Multihash, 0, len(cids))
+	for c := range cids {
+		mhs = append(mhs, c.Hash())
+	}
+
+	sort.SliceStable(mhs, func(i, j int) bool {
+		return mhs[i].String() < mhs[j].String()
+	})
+
+	log.Infow("Returning a chunk from MultihashLister", "contextId", contextIdStr, "size", len(mhs))
+
+	return provider.SliceMultihashIterator(mhs), nil
+}
+
+// NewReframeListenerWithNonceGen creates a reframe listener and initialises its state from the provided datastore.
+func New(ctx context.Context, engine provider.Interface, cidTtl time.Duration, chunkSize int, providerId string, addresses []string, ds datastore.Datastore, nonceGen func() []byte) (*ReframeListener, error) {
+	if nonceGen == nil {
+		nonceGen = defaultNonceGen
+	}
+
+	indexProvider := &ReframeListener{
+		engine:                 engine,
+		cidTtl:                 cidTtl,
+		chunkSize:              chunkSize,
+		currentChunk:           &cidsChunk{Cids: make(map[cid.Cid]struct{})},
+		listNodeByCid:          make(map[cid.Cid]*list.Element),
+		chunkByContextId:       make(map[string]*cidsChunk),
+		chunkByCid:             make(map[cid.Cid]*cidsChunk),
+		dsWrapper:              &dsWrapper{ds: ds},
+		nonceGen:               nonceGen,
+		lastSeenProviderInfo:   &peer.AddrInfo{},
+		configuredProviderInfo: nil,
```

Comment:
Could these be a single variable, and the initial value comes from configuration and gets replaces by the last valid value seen?

Reply:
Unfortunatelly no - the punch line is that it will use the one from configuration if that has been provided. I'm thinking of changing it to union of the two, hence two variables.